cpl/module_block_data.F90

```diff
@@ -0,0 +1,483 @@
+    real(kind_phys), optional, intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb
+    real(kind_phys) :: factor
+
+    ! -- begin
+    if (associated(destin_ptr) .and. associated(source_ptr)) then
+      if (slice > 0 .and. slice <= size(source_ptr, dim=2)) then
+        factor = 1._kind_phys
+        if (present(scale_factor)) factor = scale_factor
+        do ix = 1, block%blksz(block_index)
+          ib = block%index(block_index)%ii(ix)
+          jb = block%index(block_index)%jj(ix)
+          i = ib - block%isc + 1
+          j = jb - block%jsc + 1
+          destin_ptr(i,j) = factor * source_ptr(ix,slice)
+        enddo
+      end if
+    end if
+
+  end subroutine block_copy_1dslice_to_2d_r8
+
+  ! -- copy: 2D to 3D
+
+  subroutine block_copy_2d_to_3d_r8(destin_ptr, source_ptr, block, block_index, scale_factor)
+
+    ! -- arguments
+    real(ESMF_KIND_R8),        pointer    :: destin_ptr(:,:,:)
+    real(kind_phys),           pointer    :: source_ptr(:,:)
+    type (block_control_type), intent(in) :: block
+    integer,                   intent(in) :: block_index
+    real(kind_phys), optional, intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb, k
+    real(kind_phys) :: factor
+
+    ! -- begin
+    if (associated(destin_ptr) .and. associated(source_ptr)) then
+      factor = 1._kind_phys
+      if (present(scale_factor)) factor = scale_factor
+      do k = 1, size(source_ptr, dim=2)
+        do ix = 1, block%blksz(block_index)
+          ib = block%index(block_index)%ii(ix)
+          jb = block%index(block_index)%jj(ix)
+          i = ib - block%isc + 1
+          j = jb - block%jsc + 1
+          destin_ptr(i,j,k) = factor * source_ptr(ix,k)
+        enddo
+      enddo
+    end if
+
+  end subroutine block_copy_2d_to_3d_r8
+
+  ! -- copy: 2D to 2D
+
+  subroutine block_copy_2d_to_2d_r8(destin_ptr, source_ptr, block, block_index, scale_factor)
+
+    ! -- arguments
+    real(ESMF_KIND_R8),        pointer    :: destin_ptr(:,:)
+    real(kind_phys),           pointer    :: source_ptr(:,:)
+    type (block_control_type), intent(in) :: block
+    integer,                   intent(in) :: block_index
+    real(kind_phys), optional, intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb
+    real(kind_phys) :: factor
+
+    ! -- begin
+    if (associated(destin_ptr) .and. associated(source_ptr)) then
+      factor = 1._kind_phys
+      if (present(scale_factor)) factor = scale_factor
+      do ix = 1, block%blksz(block_index)
+        ib = block%index(block_index)%ii(ix)
+        jb = block%index(block_index)%jj(ix)
+        i = ib - block%isc + 1
+        j = jb - block%jsc + 1
+        destin_ptr(i,j) = factor * source_ptr(ib,jb)
+      enddo
+    end if
+
+  end subroutine block_copy_2d_to_2d_r8
+
+  subroutine block_array_copy_2d_to_2d_r8(destin_ptr, source_arr, block, block_index, scale_factor)
```

Comment:
Does `_r8` in the subroutine name refers to source or destination kind?

Reply:
The destination kind. So far the source kind is always real*4.
We can certainly add the source kind to the procedure's name.

Reply:
Is it? Looks to me source kind can be either 4 or 8 depending on how we build the model. If you want this routine to be called only for 4-byte reals then explicitly set the kind of source_arr.

Reply:
You are right, I should have been more clear. The parametric kind dependency takes care of the source argument, so we don't need to replicate more code.